br/pkg/lightning/errormanager/errormanager.go

```diff
@@ -504,8 +534,143 @@ func (em *ErrorManager) ResolveAllConflictKeys(
 	return errors.Trace(g.Wait())
 }
 
+// ReplaceConflictKeys query all conflicting rows (handle and their
+// values) from the current error report and resolve them concurrently
+// by replacing the necessary rows and reserving the others.
+func (em *ErrorManager) ReplaceConflictKeys(
+	ctx context.Context,
+	tbl tidbtbl.Table,
+	tableName string,
+	pool *utils.WorkerPool,
+	decoder *kv.TableKVDecoder,
+	fnGetLatest func(ctx context.Context, key []byte) ([]byte, error),
+	fnDeleteKey func(ctx context.Context, key []byte) error,
+) error {
+	if em.db == nil {
+		return nil
+	}
+
+	sessionOpts := encode.SessionOptions{
+		// TODO: need to find the correct value for SQLMode
+		SQLMode: mysql.ModeStrictAllTables,
+	}
+	encoder, err := kv.NewBaseKVEncoder(&encode.EncodingConfig{
+		Table:          tbl,
+		SessionOptions: sessionOpts,
+		Logger:         log.L(),
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	g, gCtx := errgroup.WithContext(ctx)
+	pool.ApplyOnErrorGroup(g, func() error {
+		// check index KV first
+		rawHandleRows, err := em.db.QueryContext(
+			gCtx, fmt.Sprintf(selectIndexConflictKeysReplace, em.schemaEscaped),
+			tableName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for rawHandleRows.Next() {
+			var rawKey []byte
+			if err := rawHandleRows.Scan(&rawKey); err != nil {
+				return errors.Trace(err)
+			}
+			em.logger.Debug("got rawKey from table",
+				zap.ByteString("rawKey", rawKey))
+
+			var value []byte
+			value, err = fnGetLatest(gCtx, rawKey)
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			var rows *sql.Rows
+			rows, err = em.db.QueryContext(
+				gCtx, fmt.Sprintf(selectConflictKeysReplaceByRawKey, em.schemaEscaped),
+				tableName, rawKey)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			for rows.Next() {
+				var indexName string
+				var rawValue []byte
+				if err := rows.Scan(&indexName, &rawValue); err != nil {
+					return errors.Trace(err)
+				}
+				em.logger.Debug("got index_name, raw_value from table",
+					zap.String("index_name", indexName),
+					zap.Binary("raw_value", rawValue),
+					zap.Binary("latest value", value))
+				if bytes.Equal(rawValue, value) {
+					continue
+				}
+				indexInfo := tbl.Meta().FindIndexByName(indexName)
+				var handle tidbkv.Handle
+				handle, err = decoder.DecodeHandleFromIndex(indexInfo, rawKey, rawValue)
+				em.logger.Debug("got handle from index",
+					zap.String("handle", handle.String()),
+					zap.Error(err))
+				rowKey := tablecodec.EncodeRowKeyWithHandle(tbl.Meta().ID, handle)
```

Comment:
is the computed `rowKey` always the same value of the `raw_handle` column in table? So we don't need to compute it again

Reply:
I have checked, and it turns out that `rowKey` indeed is always the same value of the `raw_handle` column in table, so I have changed it to use `raw_handle` directly instead of computing `rowKey`.